[FLINK-28765] Create a doc for protobuf format (#20408)
apache/flink: docs/content/docs/connectors/table/formats/protobuf.md

    </tr>
    <tr>
      <td><h5>protobuf.ignore-parse-errors</h5></td>
      <td>optional</td>
      <td>yes</td>
      <td style="word-wrap: break-word;">false</td>
      <td>Boolean</td>
      <td>Optional read flag to skip rows with parse errors instead of failing.</td>
    </tr>
    <tr>
      <td><h5>protobuf.read-default-values</h5></td>
      <td>optional</td>
      <td>yes</td>
      <td style="word-wrap: break-word;">false</td>
      <td>Boolean</td>
      <td>
          The option only works if then generated class's version is proto2. If the config is set to true, the format will read empty values as default values defined in proto file.
          If the config is set to false, the format will generate null value if the data point does not exist in binary protobuf message.
          If proto syntax is proto3, this value will be set true forcibly because proto3's standard is to use default values.
      </td>
    </tr>
    <tr>
      <td><h5>protobuf.write-null-string-literal</h5></td>
      <td>optional</td>
      <td>yes</td>
      <td style="word-wrap: break-word;">""</td>
      <td>String</td>
      <td>

>           When serializing to protobuf data, this is the optional config to specify the string literal in protobuf's array/map in case of null values.

Reviewer: ```suggestion
          When serializing to protobuf data, this is the optional config to specify the string literal in Protobuf's array/map in case of null values.
```

      <td><code>int32</code></td>
      <td></td>
    </tr>
    <tr>
      <td><code>BIGINT</code></td>
      <td><code>int64</code></td>
      <td></td>
    </tr>
    <tr>
      <td><code>FLOAT</code></td>
      <td><code>float</code></td>
      <td></td>
    </tr>
    <tr>
      <td><code>DOUBLE</code></td>
      <td><code>double</code></td>
      <td></td>
    </tr>
    <tr>
      <td><code>ARRAY</code></td>
      <td><code>repeated</code></td>
      <td>Elements cannot be null, the string default value can be specified by <code>write-null-string-literal</code></td>
    </tr>
    <tr>
      <td><code>MAP</code></td>
      <td><code>map</code></td>
      <td>Keys or values cannot be null, the string default value can be specified by <code>write-null-string-literal</code></td>
    </tr>
    <tr>
      <td><code>ROW</code></td>
      <td><code>message</code></td>
      <td></td>
    </tr>
    <tr>
      <td><code>VARCHAR / CHAR / TINYINT / SMALLINT / INTEGER / BIGINT</code></td>
      <td><code>enum</code></td>
      <td>The enum value of protobuf can be mapped to string or number of flink row accordingly.</td>
    </tr>
    </tbody>
</table>

Null Values
----------------

> As protobuf does not permit null value in map and array. We need to auto generate default value for them when converting from flink row to protobuf.

Reviewer: ```suggestion
As protobuf does not permit null values in maps and array, we need to auto-generate default values when converting from Flink Rows to Protobuf.
```

<table class="table table-bordered">
    <thead>
      <tr>
        <th class="text-left">Protobuf Data Type</th>
        <th class="text-left">Default Value</th>
      </tr>
    </thead>
    <tbody>
    <tr>
      <td>int32 / int64 / float / double</td>
      <td>0</td>
    </tr>
    <tr>
      <td>string</td>
      <td>""</td>
    </tr>
    <tr>
      <td>bool</td>
      <td>false</td>
    </tr>
    <tr>
      <td>enum</td>
      <td>first element of enum</td>
    </tr>
    <tr>
      <td>binary</td>
      <td>ByteString.EMPTY</td>
    </tr>
    <tr>
      <td>message</td>
      <td>MESSAGE.getDefaultInstance()</td>
    </tr>
    </tbody>
</table>

OneOf field
----------------

> In serialization process, there's no guarantee that the flink fields of the same one-of group only contain at most one valid value.
> When serialization, we set each field in the order of flink schema, so the field in higher position will override then field of lower position in the same one-of group.

Reviewer: ```suggestion
In the serialization process, there's no guarantee that the Flink fields of the same one-of group only contain at most one valid value.
When serializing, each field is set in the order of Flink schema, so the field in the higher position will override the field in lower position in the same one-of group.
```

The Protocol Buffers [Protobuf](https://developers.google.com/protocol-buffers) format allows you to read and write Protobuf data, based on Protobuf generated classes.

Dependencies
------------

{{< sql_download_table "protobuf" >}}

How to create a table with Protobuf format
----------------

Here is an example to create a table using the Kafka connector and Protobuf format.

Below is the proto definition file.

```
syntax = "proto2";
package com.example;
option java_package = "com.example";
option java_multiple_files = true;

message SimpleTest {
    optional int64 uid = 1;
    optional string name = 2;
    optional int32 category_type = 3;
    optional bytes content = 4;
    optional double price = 5;
    map<int64, InnerMessageTest> value_map = 6;
    repeated  InnerMessageTest value_arr = 7;
    optional Corpus corpus_int = 8; 
    optional Corpus corpus_str = 9; 
    
    message InnerMessageTest{
          optional int64 v1 =1;
          optional int32 v2 =2;
    }
    
    enum Corpus {
        UNIVERSAL = 0;
        WEB = 1;
        IMAGES = 2;
        LOCAL = 3;
        NEWS = 4;
        PRODUCTS = 5;
        VIDEO = 7;
      }
}
```

> Use `protoc` command to generate protobuf java sources and put the sources in your runtime classpath.

Reviewer 2: ```suggestion
1. Use [`protoc`](https://developers.google.com/protocol-buffers/docs/javatutorial#compiling-your-protocol-buffers) command to compile the `.proto` file to java classes
2. Then compile and package the classes 
3. Finally you should provide the `jar` in your classpath, e.g. pass it using `-j` in <a href="{{< ref "docs/dev/table/sqlClient" >}}">sql-client</a>
```
